File: src/engine/commands/givecommand.cpp
#include "givecommand.h"


#define super Command

GiveCommand::GiveCommand(QObject *parent) :
    super(parent) {

    setDescription("Give an item from your inventory to another character.\n"
                   "\n"
                   "Examples: give stick earl, give $5 to beggar");
}

GiveCommand::~GiveCommand() {
}

void GiveCommand::execute(Player *player, const QString &command) {

    super::prepareExecute(player, command);

    if (!assertWordsLeft("Give what?")) {
        return;
    }

    GameObjectPtrList items = takeObjects(player->inventory());
    if (!requireSome(items, "You don't have that.")) {
        return;
    }

    takeWord("to", IfNotLast);

    GameObjectPtrList recipients = takeObjects(currentRoom()->characters());
    if (!requireSome(recipients, "That recipient is not here.")) {
        return;
    }

    Character *recipient = recipients[0].cast<Character *>();
    QString recipientName = recipient->definiteName(currentRoom()->characters());

    GameObjectPtrList givenItems;
    for (const GameObjectPtr &itemPtr : items) {
        Item *item = itemPtr.cast<Item *>();
        if (recipient->inventoryWeight() + item->weight() > recipient->maxInventoryWeight()) {
            send("%1 is too heavy for %2.",
                 item->definiteName(player->inventory(), Capitalized), recipientName);
        } else if (!recipient->invokeTrigger("onreceive", player, itemPtr)) {
            continue;
        } else {
            recipient->addInventoryItem(itemPtr);
            givenItems.append(itemPtr);
        }
    }

    QString description = givenItems.joinFancy();

    for (const GameObjectPtr &itemPtr : givenItems) {
        player->removeInventoryItem(itemPtr);
    }

    send("You give %1 to %2.", description, recipientName);
    recipient->send(QString("%1 gives you %2.").arg(player->name(), description));

    GameObjectPtrList others = currentRoom()->players();
    others.removeOne(player);
    others.removeOne(recipient);
    others.send(QString("%1 gives %2 to %3.").arg(player->name(), description, recipientName));
}

File: data/commands/givecommand.js
function GiveCommand() {
    Command.call(this);

    this.setDescription("Give an item from your inventory to another character.\n" +
                        "\n" +
                        "Examples: give stick to earl");
}

GiveCommand.prototype = new Command();
GiveCommand.prototype.constructor = GiveCommand;

GiveCommand.prototype.execute = function(player, command) {

    this.prepareExecute(player, command);

    if (!this.assertWordsLeft("Give what?")) {
        return;
    }

    var items = this.takeObjects(player.inventory);
    if (!this.requireSome(items, "You don't have that.")) {
        return;
    }

    this.takeWord("to", Options.IfNotLast);

    var recipient = this.takeObject(this.currentRoom.characters);
    if (!this.requireSome(recipient, "That recipient is not here.")) {
        return;
    }

    var recipientName = recipient.definiteName(this.currentRoom.characters);

    var givenItems = [];
    for (var i = 0, length = items.length; i < length; i++) {
        var item = items[i];
        if (recipient.inventoryWeight() + item.weight > recipient.maxInventoryWeight()) {
            this.send("%1 is too heavy for %2.",
                      item.definiteName(player.inventory, Options.Capitalized), recipientName);
        } else if (!recipient.invokeTrigger("onreceive", player, item)) {
            continue;
        } else {
            recipient.addInventoryItem(item);
            player.removeInventoryItem(item);
            givenItems.append(item);
        }
    }

    var description = givenItems.joinFancy();

    this.send("You give %1 to %2.", description, recipientName);
    recipient.send("%1 gives you %2.".arg(player.name, description));

    var others = this.currentRoom.characters;
    others.removeOne(player);
    others.removeOne(recipient);
    others.send("%1 gives %2 to %3.".arg(player.name, description, recipientName));
};

CommandRegistry.registerCommand("give", new GiveCommand());
